### ISSUE TITLE
Bug: [no-unnecessary-condition] false positive in record index

### ISSUE THREAD

**Issue by PowerKiKi** [typescript-eslint/typescript-eslint#11666]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=GYVwdgxgLglg9mABFApgZygCgIYC5EYBOMYA5gJT5EmmIDeAUIohAhogLbYAO%2BASilaEAJgB5qZADSIA5ABsEpAPphsHFDIB8iALz0mzRITghU%2BeYpVqNkgwF8A3AwMxgiTF24BtbAF1y%2BoZGKFAghEiePr5OzHbOzIQhYUgywijA2CByUIgAbthyIBpOcUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
function test(a: string): string {
  const map: Record<string, 'long_name'> = {
    route: 'long_name',
  };

  if (map[a]) {
    return map[a];
  }

  return 'default value';
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expect to see no error. Because the key `a` may or may not exist in `map`, then the `if` must exists, and should absolutely not be removed.

### Actual Result

The following error is incorrectly reported:

```
Unnecessary conditional, value is always truthy. 6:7 - 6:13
```

### Additional Info

Interestingly enough this work as expected [in a tsx file](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=GYVwdgxgLglg9mABFApgZygCgIYC5EYBOMYA5gJT5EmmIDeAUIohAhogLbYAO%2BASilaEAJgB5qZADSIA5ABsEpAPphsHFDIB8iALz0mzRITghU%2BeYpVqNkgwF8A3AwMxgiTF24BtbAF1y%2BoZGKFAghEiePr5OzHbOzIQhYUgywijA2CByUIgAbthyIBpOcUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false).

It also work if `map` type is modified to [have values of `string`](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=GYVwdgxgLglg9mABFApgZygCgIYC5EYBOMYA5gJT5EmmIDeAUIohAhogLbYAO%2BASilaEAJgB5qZADQEoxMgD5EAXnpNmiQnBCp8AcgA2CUgH0w2Dil2S1AXwDcDNTGCJMXbgG1sAXXKr1GihQIIRI7l7eDsw2jsyEQSFIusIowNgg%2BlCIAG7Y%2BiCWDjFAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false), instead of `'long_name'`.

**Comment by kirkwaiblinger** [comment]
Hi @PowerKiKi!

This is one of the classic [limitations of the rule](https://typescript-eslint.io/rules/no-unnecessary-condition/#limitations).

> this works in a tsx file

Your link seems to just not have the rule enabled... see [link with rule enabled](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=GYVwdgxgLglg9mABFApgZygCgIYC5EYBOMYA5gJT5EmmIDeAUIohAhogLbYAO%2BASilaEAJgB5qZADSIA5ABsEpAPphsHFDIB8iALz0mzRITghU%2BeYpVqNkgwF8A3AwMxgiTF24BtbAF1y%2BoZGKFAghEiePr5OzHbOzIQhYUgywijA2CByUIgAbthyIBpOcUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRQA7iKaRwYAL4hNQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

> it also works with `string`

This is because `'long_name'` is never falsy, whereas `string` might be `''` which is falsy. If you include the possibility of the empty string, the report goes away again ([link](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=GYVwdgxgLglg9mABFApgZygCgIYC5EYBOMYA5gJT5EmmIDeAUIohAhogLbYAO%2BASilaEAJgB5qZADSIA5ABsEpAPphsHFDMQAfWTIB8iALz0mzRITghU%2BeYpVqNk0wF8A3A1MxgiTF24BtbABdchMzcxQoEEIkP0Cg92ZnD2ZCSOikGWEUYGwQOShEADdsORANd2SgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRQA7iKaRwYAL4hNQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)). You can observe the same pattern of behavior as well with `Record<string, number>` vs `Record<string, 42>` vs `Record<string, 42 | 0>`.

---

I'd say that's the end of the story more or less, but I think there is supposed to be a small amount of special-casing to make the rule avoid reporting in array/index access situations like this, even without `noUncheckedIndexedAccess` enabled, so I'd want to double check the expected behavior before closing.

**Comment by kirkwaiblinger** [comment]
Looks like we specifically test to not special case this... but it seems worth considering while we're in the business of special-casing arrays too?


FWIW I found this case because I prototyped adding index access detection to the current array access detection and it seemed pretty straightforward.

https://github.com/typescript-eslint/typescript-eslint/blob/0cff7bf85a3d9bd6dccddce53124ec6944eedfe3/packages/eslint-plugin/tests/rules/no-unnecessary-condition.test.ts#L1912-L1922

@typescript-eslint/triage-team

**Comment by PowerKiKi** [comment]
Thank you for pointing me to the limitations I was not aware of. I understand the behavior of TypeScript to be optimistic by default (to reduce boilerplate in real world cases), But I feel typescript-eslint should, on the contrary, always be pessimistic to avoid (possibly) breaking code with auto fixes.

For the time being I keep this rule disabled, because it is too dangerous for my codebase.

**Comment by JoshuaKGoldberg** [comment]
The fewer special cases we have the better IMO. My vote is to wontfix as by design for TypeScript. Even though, yes, this is an irksome situation with `Record`s.

**Comment by PowerKiKi** [comment]
wontfix-ing this issue is fine by me. It is a bit unfortunate, since it prevent me from auto-fixing other unnecessary condition, but I understand that this project resources might be better invested elsewhere than juggling with special-cases left and right.

**Comment by JoshuaKGoldberg** [comment]
Closing out as an unfortunate wontfix. Thanks for the discussion, all!
